README.md

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Signed
by](https://img.shields.io/badge/Keybase-Verified-brightgreen.svg)](https://keybase.io/hrbrmstr)
![Signed commit
%](https://img.shields.io/badge/Signed_Commits-0%25-lightgrey.svg)
[![R-CMD-check](https://github.com/hrbrmstr/ggsolar/workflows/R-CMD-check/badge.svg)](https://github.com/hrbrmstr/ggsolar/actions?query=workflow%3AR-CMD-check)
![Minimal R
Version](https://img.shields.io/badge/R%3E%3D-3.6.0-blue.svg)
![License](https://img.shields.io/badge/License-MIT-blue.svg)

# ggsolar

Generate Solar System Plots

## Description

Generate “solar system” plots using everyone’s favorite plotting
package.

## What’s Inside The Tin

The following functions are implemented:

- `generate_orbits`: Generate concentric circular orbits for “n” planets
- `generate_random_planets`: Generate a vector of n believable planetary
  names using hidden Markov model.
- `plot_orbits`: Return a skeleton of a ggplot2 plot of solar system
  plot
- `randomize_planet_positions`: Generate random planetary positions
  given a data frame of orbits
- `scaffold_planet_plot`: Get a boilerplate to work from
- `sol_planets`: Our Solar System
- `theme_enhance_solar`: Provides basic cleanup defaults for solar
  system plots

## Installation

``` r
remotes::install_github("hrbrmstr/ggsolar")
```

NOTE: To use the ‘remotes’ install options you will need to have the
[{remotes} package](https://github.com/r-lib/remotes) installed.

## Usage

``` r
library(ggsolar)
library(ggplot2)

# current version
packageVersion("ggsolar")
## [1] '0.1.1'
```

## Sol

``` r
sol_orbits <- generate_orbits(sol_planets)

set.seed(1323) # this produced decent placements

placed_planets <- randomize_planet_positions(sol_orbits)

plot_orbits(
  orbits = sol_orbits, 
  planet_positions = placed_planets,
  label_planets = TRUE,
  label_family = hrbrthemes::font_es_bold
) +
  hrbrthemes::theme_ipsum_es(grid="") +
  coord_equal() +
  labs(
    title = "Sol",
    caption = "Pluto is 100% a planet"
  ) +
  theme_enhance_solar()
```

<img src="man/figures/README-sol-1.png" width="576" />

## Non-uniform Radii

We’ll plot Sol’s system with the square root of the actual radius AU’s
(which are `0.39`, `0.72`, `1.00`, `1.52`, `5.20`, `9.58`, `19.18`,
`30.07`, `39.48`):

``` r
sol_dist <- sqrt(c(0.39, 0.72, 1.00, 1.52, 5.20, 9.58, 19.18, 30.07, 39.48))
sol_orbits <- generate_orbits(sol_planets, radii = sol_dist, num_polygon_points = 1000)

set.seed(42)
sol_placements <- randomize_planet_positions(sol_orbits)

plot_orbits(
  orbits = sol_orbits, 
  planet_positions = sol_placements,
  label_planets = TRUE,
  label_family = hrbrthemes::font_es_bold
) +
  hrbrthemes::theme_ipsum_es(grid="") +
  coord_equal() +
  labs(
    title = "Sol System (Orbits AU Sqrt.)",
    subtitle = "Non-uniform Radii",
    caption = "Pluto is 100% a planet"
  ) +
  theme_enhance_solar()
```

<img src="man/figures/README-sol-sqrt-1.png" width="672" />

## Custom Planet Placements!

We’ll plot Sol’s system with the square root of the actual radius AU’s
(which are `0.39`, `0.72`, `1.00`, `1.52`, `5.20`, `9.58`, `19.18`,
`30.07`, `39.48`):

``` r
sol_dist <- sqrt(c(0.39, 0.72, 1.00, 1.52, 5.20, 9.58, 19.18, 30.07, 39.48))
sol_orbits <- generate_orbits(sol_planets, radii = sol_dist, num_polygon_points = 1000)

my_pos <- function(n) {
  seq(from = 0, to = 1, length.out = 9)
}

sol_placements <- randomize_planet_positions(sol_orbits, randomizer = my_pos)

plot_orbits(
  orbits = sol_orbits, 
  planet_positions = sol_placements,
  label_planets = TRUE,
  label_family = hrbrthemes::font_es_bold
) +
  hrbrthemes::theme_ipsum_es(grid="") +
  coord_equal() +
  labs(
    title = "Sol System (Orbits AU Sqrt.)",
    subtitle = "Custom placements",
    caption = "Pluto is 100% a planet"
  ) +
  theme_enhance_solar()
```

<img src="man/figures/README-sol-precise-sqrt-1.png" width="672" />

## Rando!

``` r
set.seed(42)
(rando_planets <- generate_random_planets(12))
##  [1] "Ososi"       "Morlambtune" "Obli VI"     "Lantitan"    "Tanthos"     "Onhy"        "Lesbomor C"  "Radido"     
##  [9] "Lesstel"     "Altan prime" "Lasdre"      "Brenta"

rando_orbits <- generate_orbits(rando_planets)

set.seed(123) # this produced decent placements

placed_planets <- randomize_planet_positions(rando_orbits)

plot_orbits(
  orbits = rando_orbits, 
  planet_positions = placed_planets,
  label_planets = TRUE,
  label_family = hrbrthemes::font_es_bold
) +
  hrbrthemes::theme_ipsum_es(grid="") +
  coord_equal() +
  labs(
    title = "Rando System"
  ) +
  theme_enhance_solar()
```

<img src="man/figures/README-rando-1.png" width="576" />

## ggsolar Metrics

| Lang | \# Files |  (%) | LoC |  (%) | Blank lines |  (%) | \# Lines |  (%) |
|:-----|---------:|-----:|----:|-----:|------------:|-----:|---------:|-----:|
| R    |        8 | 0.36 | 215 | 0.32 |          47 | 0.25 |      154 | 0.39 |
| Rmd  |        1 | 0.05 |  80 | 0.12 |          38 | 0.20 |       42 | 0.11 |
| YAML |        2 | 0.09 |  38 | 0.06 |          10 | 0.05 |        2 | 0.01 |
| SUM  |       11 | 0.50 | 333 | 0.50 |          95 | 0.50 |      198 | 0.50 |

clock Package Metrics for ggsolar

## Code of Conduct

Please note that this project is released with a Contributor Code of
Conduct. By participating in this project you agree to abide by its
terms.


NEWS.md
0.1.1
* Switched over to {ggforce} and `geom_circle()`

0.1.0 
* Initial release


tests/tinytest.R

if ( requireNamespace("tinytest", quietly=TRUE) ){
  tinytest::test_package("ggsolar")
}



inst/tinytest/test_ggsolar.R
library(ggsolar)

shld_be <- c("Lesra VIII", "Tinus proxima", "Nigplu D", "Nusam", "Gussiania",
             "Aenma", "Depe D", "Caty proxima", "Magnus", "Cosamlos")

set.seed(1)
planets <- generate_random_planets(10)

expect_equivalent(shld_be, planets)

orbs <- generate_orbits(planets, num_polygon_points = 5)
expect_equivalent(orbs, structure(list(planet = shld_be,
radius = 1:10, num_polygon_points = c(5, 5, 5, 5, 5, 5, 5, 5, 5, 5)),
class = "data.frame", row.names = c(NA, -10L
)))
set.seed(3)
ppos <- randomize_planet_positions(orbs)

expect_equivalent(ppos, structure(list(planet = shld_be,
angle = c(1.05583604929519, 6.07377517395879, 4.41866412626915,
5.05921544626883, 7.78311011310412, 8.7975198400928, 6.78309502563394,
8.85103219869237, 11.6292301564148, 12.9645597060423), x = c(0.492500278444618,
1.95630741627465, -0.868558655307667, 1.35965981757038, 0.354061035904789,
-4.85783899670659, 6.14338089114735, -6.71898992310528, 5.32885371204113,
9.21764684274087), y = c(0.87031228632714, -0.41576591133572,
     -2.87151630019579, -3.76182471421576, 4.98744832382793, 3.52156219341313,
     3.35542414998248, 4.34225453114044, -7.25281449615702, 3.87749747678955
)), class = "data.frame", row.names = c(NA, -10L)))



R/aaa.R
utils::globalVariables(
  c("x", "y", "planet", "radius")
)

#' Our Solar System
#'
#' Character vector of the planet names of Sol
#'
#' @note To hades with "science"! Pluto is 100%% a bona fine planet!
#' @docType data
#' @export
sol_planets <- c("Mercury", "Venus", "Earth", "Mars", "Jupyter", "Saturn", "Uranus", "Neptune", "Pluto")

c("cen-tau-ri", "pro-xi-ma", "al-pha", "pro-xi-mus", "ter-ra",
  "sa-turn", "ju-pi-ter", "mer-cu-ry", "ve-nus", "u-ra-nus", "mars",
  "nep-tune", "plu-to", "ni-bi-ru", "tra-pist", "zeus", "aph-ro-dite",
  "a-res", "heph-aes-tus", "ty-phoon", "da-scher", "en-dor", "ka-mi-no",
  "qua-sar", "no-va", "om-ni-sen", "dune", "a-ra-kis", "gla-tius",
  "co-do-mi-ni-um", "dor-sai", "o-si-mio", "he-li-os", "leo", "hein-lein",
  "a-ca-mar", "ach-ra-dy", "a-di-ge-on", "a-las-tria", "al-drea",
  "von-tria", "al-di-bran", "al-ge-ron", "ach-eon", "a-lon-dra",
  "cyg-nus", "a-ri-done", "a-ri-da-ni", "lep-tus", "am-leth", "am-di-vian",
  "om-ri-con", "ma-jo-ris", "mi-no-ris", "leo-nis", "an-do-ria",
  "an-go-sia", "ar-ge-lius", "ar-gus", "per-cii", "pro-me-thi-us",
  "the-os", "ar-ret", "ar-va-da", "au-re-lia", "ba-lon-si", "bar-son",
  "ba-ri-di-on", "ben-zar", "ben-thos", "kup-sic", "bo-kara", "bo-la-rus",
  "bo-ra-dis", "bos-lic", "bren-ta-lia", "bri-toid", "he-lia",
  "cal-do-nia", "cal-dos", "cal-leb", "ca-li-os", "ca-mus", "ca-no-pius",
  "car-da-sia", "cas-tal", "ca-tu-lia", "ce-ti", "cor-van", "co-ri-dan",
  "cyg-nia", "cyg-net", "da-ka-ra", "da-lan", "da-nu-la", "oxi-lia",
  "dav-los", "de-cos", "de-des-tris", "cen-tris", "cat-los", "der-bia",
  "der-mi-nia", "de-ni-us", "nig-mus", "stag-ni-us", "spho-li-us",
  "shpo-rio", "dre-on", "val-les", "al-pes", "al-pus", "xar-ria",
  "cor-do-via", "met-ne-ria", "cos-ma-dov", "ju-ne-ria", "del-ta",
  "o-rio-na", "ci-ca-da", "am-ne-li-um", "noe-li-um", "au-ra",
  "au-ro-ra", "em-bo-ria", "phan-tax", "ob-lox", "ob-li-via", "za-thu-ra",
  "ar-chea", "cen-tus", "ep-si-lon", "au-pho-ria", "au-ro-ria",
  "ma-ri-dia", "po-tel-mia", "hy-pho-ria", "ti-tan", "la-vi-tan",
  "no-mi-cron", "mag-nus", "kep-lar", "del-phi", "nim-bus", "am-phus",
  "mor-phus", "lamb-dus", "dus-to-pia", "mo-pi-um", "mo-li-um",
  "pho-ri-um", "pho-ri-oso", "oso-li-um", "xe-li-um", "so-li-sia",
  "so-lus", "ap-po-lus", "pho-bos", "chro-nos", "as-tos", "seg-mos",
  "co-los", "dei-mos", "cog-nus", "sig-mus", "co-ro-li-os", "pe-ri-li-os",
  "sep-tus", "mes-mus", "op-tus", "ne-bu-lus", "hec-tus", "me-gus",
  "gi-gus", "te-rus", "pe-tus", "lu-mus", "pri-mus", "op-ti-mus",
  "ori-gus", "xe-lo-phus", "meg-lo-vus", "al-pha-ni-us", "es-la-var",
  "cel-les-ti-us", "stel-le-ra-sus", "ra-le-os", "pa-go-ya", "pul-sar"
) -> planet_fragments


R/theme-enhance-solar.R
#' Provides basic cleanup defaults for solar system plots
#'
#' @param bg,fg background/foreground colors
#' @export
theme_enhance_solar <- function(bg = "black", fg = "white") {

  theme(
    text = element_text(color = fg),
    plot.title = element_text(color = fg),
    plot.subtitle = element_text(color = fg),
    plot.caption = element_text(color = fg),
    plot.background = element_rect(fill = bg, color = NA),
    panel.background = element_rect(fill = bg, color = NA),
    panel.grid.major.x = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    axis.text.x.bottom = element_blank(),
    axis.text.y.left = element_blank()
  )

}

R/generate-random-planets.R
#' Generate a vector of `n` believable planetary names using hidden Markov model.
#'
#' @param num_planets how many planets?
#' @return character vector of planet names
#' @references Riffed from [planet-name-generator](https://github.com/sayamqazi/planet-name-generator).
#' @export
#' @examples
#' library(ggplot2)
#'
#' generate_random_planets(15)
#'
#' set.seed(42)
#' rando_planets <- generate_random_planets(12)
#'
#' rando_orbits <- generate_orbits(rando_planets)
#'
#' set.seed(123) # this produced a decent placements
#'
#' placed_planets <- randomize_planet_positions(rando_orbits)
#'
#' plot_orbits(
#'   orbits = rando_orbits,
#'   planet_positions = placed_planets,
#'   label_planets = TRUE,
#' ) +
#'   ggplot2::theme_void() +
#'   ggplot2::coord_equal() +
#'   ggplot2::labs(
#'     title = "Rando System"
#'   ) +
#'   theme_enhance_solar()

generate_random_planets <- function(num_planets = 20) {

  planets <- planet_fragments

  syllables <- stringi::stri_split_fixed(planets, "-", simplify = TRUE)

  total_syllables <- length(syllables)

  syllables <- unique(syllables)

  div_index <- length(syllables) / total_syllables

  size <- length(syllables) + 1

  freq <- matrix(0, nrow = size, ncol = size)

  for (p in planets) {
    lex <- stringi::stri_split_fixed(p, "-", simplify = TRUE)
    lex_indices <- match(lex, syllables) - 1

    if (length(lex_indices) > 1) {
      row_indices <- lex_indices[-length(lex_indices)]
      col_indices <- lex_indices[-1]
      freq[cbind(row_indices, col_indices)] <- freq[cbind(row_indices, col_indices)] + 1
    }

    last_syllable_index <- lex_indices[length(lex_indices)]
    freq[last_syllable_index, size - 1] <- freq[last_syllable_index, size - 1] + 1
  }

  max_names <- num_planets
  num_names <- 0
  planet_name <- ""

  c("prime", "",
    "B", "",
    "alpha", "",
    'proxima', "",
    "IV", "",
    "V", "",
    "C", "",
    "VI", "",
    "VII", "",
    "VIII", "",
    "X", "",
    "IX", "",
    "D", "",
    "", "") -> suffixes

  generated_planet_names <- c()

  while (num_names < max_names) {

    length <- sample(2:3, 1)
    initial <- sample(0:(size - 2), 1)

    planet_name <- ""
    while (length > 0) {
      while (all(freq[initial + 1, ] != 1)) {
        initial <- sample(0:(size - 2), 1)
      }

      planet_name <- paste0(planet_name, syllables[initial + 1])
      initial <- which(freq[initial + 1, ] == 1)[1] - 1
      length <- length - 1
    }

    suffix_index <- sample(1:length(suffixes), 1)
    planet_name <- paste(stringi::stri_trans_totitle(planet_name), suffixes[suffix_index])
    generated_planet_names <- append(generated_planet_names, planet_name)

    num_names <- num_names + 1
  }

  stringi::stri_trim_both(generated_planet_names)

}


R/randomize-planet-positions.R
#' Generate random planetary positions given a data frame of orbits
#'
#' You can use this after calling [generate_orbits()] to compute a
#' random position along the orbit circle for each planet
#'
#' @param orbits (df) output of [generate_orbits()]
#' @param randomizer (ƒ) if not the default `runif`, a function that can take
#'        a parameter `n`, the number of orbits, and return a numeric vector of
#'        values between 0 and 1 that will position the point where you want it
#'        to be on the orbit circle.
#' @return (data.fame) with the `planet` name, computed angle, and `x`/`y` coord of the planet
#' @export
#' @examples
#' sol_orbits <- generate_orbits(sol_planets)
#' randomize_planet_positions(sol_orbits)
randomize_planet_positions <- function(orbits, randomizer=runif) {

  angles <- (seq_along(orbits$radius)-1) + randomizer(nrow(orbits)) * 2 * pi

  data.frame(
    planet = orbits$planet,
    angle = angles,
    x = orbits$radius * cos(angles),
    y = orbits$radius * sin(angles)
  )

}



R/plot-orbits.R
#' Return a skeleton of a ggplot2 plot of solar system plot
#'
#' THIS JUST A LAZY HELPER FUNCTION UNTIL I HAVE TIME TO WRITE GEOMS
#'
#' If you need to be able to vary planet size/color/fill/etc by the
#' planet name, or you want different orbit aesthetics mapped by planet,
#' you'll need, for now, to run [scaffold_planet_plot()] which will place
#' a textual copy of this function to stdout that you can copy and
#' use at-will.
#'
#' Given a data frame built with [generate_orbits()] and another
#' data frame of planet positions, theoretically generated by
#' [randomize_planet_positions()], return a ggplot2 skeleton with —
#' at a minimum — a polygon geom and point geom. A text geom can
#' optionally be added, but only naive positioning of labels is performed.
#'
#' @param orbits (df)
#' @param planet_positions (df)
#' @param label_planets (bool) if `TRUE`, then planet labels will be
#'        naively placed to the right of each planet point.
#' @param orbit_fill,orbit_color,orbit_linewidth defaults
#' @param planet_size,planet_shape,planet_stroke,planet_color,planet_fill defaults
#' @param label_color,label_size,label_family,label_hjust defaults (if `label_planets` is `TRUE`)
#' @return ggplot2 object
#' @export
#' @examples
#' library(ggplot2)
#'
#' sol_orbits <- generate_orbits(sol_planets)
#'
#' set.seed(1323)
#'
#' placed_planets <- randomize_planet_positions(sol_orbits)
#'
#' plot_orbits(
#'   orbits = sol_orbits,
#'   planet_positions = placed_planets,
#'   label_planets = TRUE,
#' ) +
#'   ggplot2::theme_void() +
#'   ggplot2::coord_equal() +
#'   ggplot2::labs(
#'     title = "Sol",
#'     caption = "Pluto is 100% a planet"
#'   ) +
#'   theme_enhance_solar()
plot_orbits <- function (orbits,
                         planet_positions,
                         label_planets = FALSE,
                         orbit_fill = NA,
                         orbit_color = "white",
                         orbit_linewidth = 0.125/2,
                         planet_size = 3,
                         planet_shape = 21,
                         planet_stroke = 0.5,
                         planet_color = "#fea100",
                         planet_fill = ggplot2::alpha("#fea100", 1/4),
                         label_color = "white",
                         label_size = 3,
                         label_family = "sans",
                         label_hjust = 0
) {

  gg <- ggplot2::ggplot()

  gg + geom_circle(
    data = orbits,
    aes(
      x0 = 0,
      y0 = 0,
      r = radius
    ),
    color = orbit_color,
    fill = orbit_fill,
    color = alpha("white", 1/2),
    size = orbit_linewidth,
  ) -> gg

  gg + ggplot2::geom_point(
    data = planet_positions,
    ggplot2::aes(x, y, group=planet),
    color = planet_color,
    fill = planet_fill,
    size = planet_size,
    shape = planet_shape,
    stroke = planet_stroke
  ) -> gg


  if (label_planets) {
    gg +  ggplot2::geom_text(
      data = planet_positions,
      ggplot2::aes(x, y, group=planet, label = sprintf("    %s", planet)),
      color = label_color,
      size = label_size,
      family = label_family,
      hjust = label_hjust
    ) -> gg
  }

  gg

}


#' Get a boilerplate to work from
#'
#' @export
scaffold_planet_plot <- function() {

  cat(
    '# Use your own data/seed!!
sol_dist <- sqrt(c(0.39, 0.72, 1.00, 1.52, 5.20, 9.58, 19.18, 30.07, 39.48))
sol_orbits <- generate_orbits(sol_planets, radii = sol_dist, num_polygon_points = 1000)

set.seed(42)
sol_placements <- randomize_planet_positions(sol_orbits)

ggplot() +
  geom_circle(
    data = sol_orbits,
    aes(
      x0 = 0,
      y0 = 0,
      r = radius
    ),
    color = alpha("white", 1/2),
    size = 0.125
  ) +
  geom_point(
    data = sol_placements,
    aes(x, y),
    color = "orange",
    fill = alpha("orange", 1/2),
    size = 3,
    shape = 21,
    stroke = 1
  ) +
  coord_fixed() +
  theme_void(grid="") +
  theme_enhance_solar()
')

}
